src/modules/flow/flower-power/flower-power.c

```diff
@@ -0,0 +1,709 @@
+    SOL_NULL_CHECK(opts->client_id, -ENOMEM);
+    mdata->client_secret = strdup(opts->client_secret);
+    SOL_NULL_CHECK(opts->client_secret, -ENOMEM);
+
+    mdata->node = node;
+
+    return 0;
+}
+
+static void
+http_get_close(struct sol_flow_node *node, void *data)
+{
+    struct http_get_data *mdata = data;
+
+    free(mdata->client_id);
+    /* Zero memory used to store client secret and password */
+    if (mdata->client_secret) {
+        memset(mdata->client_secret, 0, strlen(mdata->client_secret) + 1);
+        free(mdata->client_secret);
+    }
+    free(mdata->username);
+    if (mdata->password) {
+        memset(mdata->password, 0, strlen(mdata->password) + 1);
+        free(mdata->password);
+    }
+
+    /* FIXME: Cancel pending connections. Need HTTP API. */
+}
+
+#define RESPONSE_CHECK_API(response_, mdata_) \
+    do { \
+        if (unlikely(!response_)) { \
+            sol_flow_send_error_packet(mdata_->node, EINVAL, \
+                "Error while reaching service."); \
+            return; \
+        } \
+        if (unlikely(response_->api_version != SOL_HTTP_RESPONSE_API_VERSION)) { \
+            SOL_ERR("Unexpected API version (response is %u, expected %u)", \
+                response->api_version, SOL_HTTP_RESPONSE_API_VERSION); \
+            return; \
+        } \
+    } while (0)
+
+#define BASE_URL "https://apiflowerpower.parrot.com/"
+#define STATUS_URL BASE_URL "sensor_data/v4/garden_locations_status"
+#define AUTH_URL BASE_URL "user/v1/authenticate"
+#define AUTH_START "Bearer "
+
+static void
+generate_token_cb(void *data, struct sol_http_response *response)
+{
+    struct http_get_data *mdata = data;
+    struct sol_json_scanner scanner;
+    struct sol_json_token token, key, value;
+    enum sol_json_loop_reason reason;
+
+    RESPONSE_CHECK_API(response, mdata);
+
+    if (response->response_code != 200) {
+        sol_flow_send_error_packet(mdata->node, EINVAL,
+            "Service returned unexpected response code: %d for request %s",
+            response->response_code, AUTH_URL);
+        return;
+    }
+
+    if (!response->content.used) {
+        sol_flow_send_error_packet(mdata->node, ENOKEY,
+            "Empty response from server for request %s", AUTH_URL);
+        return;
+    }
+
+    sol_json_scanner_init(&scanner, response->content.data,
+        response->content.used);
+    SOL_JSON_SCANNER_OBJECT_LOOP (&scanner, &token, &key, &value, reason) {
+        if (!sol_json_token_str_eq(&key, "access_token",
+                strlen("access_token")))
+            continue;
+        else {
+            size_t auth_len = strlen(AUTH_START);
+            /* value is between double quotes */
+            size_t token_size = value.end - value.start + auth_len - 1;
+
+            free(mdata->token);
+            mdata->token = malloc(token_size);
+            SOL_NULL_CHECK(mdata->token);
+
+            strcpy(mdata->token, AUTH_START);
+            memcpy(mdata->token + auth_len, value.start + 1,
```

Comment:
Yes, but you received this from the network and you're blindly accepting that the contents are within the spec.